drivers/gpu/drm/panel/panel-k6-38-0e-0b-dsc.c

```diff
@@ -18,415 +18,415 @@
-	dsi->mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
-			  MIPI_DSI_CLOCK_NON_CONTINUOUS;
-
-	drm_panel_init(&ctx->panel, dev, &k6_38_0e_0b_dsc_panel_funcs,
-		       DRM_MODE_CONNECTOR_DSI);
-	ctx->panel.prepare_prev_first = true;
-
-	ctx->panel.backlight = k6_38_0e_0b_dsc_create_backlight(dsi);
-	if (IS_ERR(ctx->panel.backlight))
-		return dev_err_probe(dev, PTR_ERR(ctx->panel.backlight),
-				     "Failed to create backlight\n");
-
-	drm_panel_add(&ctx->panel);
-
-	/* This panel only supports DSC; unconditionally enable it */
-	dsi->dsc = &ctx->dsc;
-
-	ctx->dsc.dsc_version_major = 1;
-	ctx->dsc.dsc_version_minor = 1;
-
-	/* TODO: Pass slice_per_pkt = 2 */
-	ctx->dsc.slice_height = 40;
-	ctx->dsc.slice_width = 540;
-	/*
-	 * TODO: hdisplay should be read from the selected mode once
-	 * it is passed back to drm_panel (in prepare?)
-	 */
-	WARN_ON(1080 % ctx->dsc.slice_width);
-	ctx->dsc.slice_count = 1080 / ctx->dsc.slice_width;
-	ctx->dsc.bits_per_component = 8;
-	ctx->dsc.bits_per_pixel = 8 << 4; /* 4 fractional bits */
-	ctx->dsc.block_pred_enable = true;
-
-	ret = mipi_dsi_attach(dsi);
-	if (ret < 0) {
-		dev_err(dev, "Failed to attach to DSI host: %d\n", ret);
-		drm_panel_remove(&ctx->panel);
-		return ret;
-	}
-
-	return 0;
+    struct device* dev = &dsi->dev;
+    struct k6_38_0e_0b_dsc* ctx;
+    int ret;
+
+    ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
+    if (!ctx)
+        return -ENOMEM;
+
+    ctx->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_HIGH);
+    if (IS_ERR(ctx->reset_gpio))
+        return dev_err_probe(dev, PTR_ERR(ctx->reset_gpio),
+            "Failed to get reset-gpios\n");
+
+    ctx->dsi = dsi;
+    mipi_dsi_set_drvdata(dsi, ctx);
+
+    dsi->lanes = 4;
+    dsi->format = MIPI_DSI_FMT_RGB888;
+    dsi->mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_CLOCK_NON_CONTINUOUS;
+
+    drm_panel_init(&ctx->panel, dev, &k6_38_0e_0b_dsc_panel_funcs,
+        DRM_MODE_CONNECTOR_DSI);
+    ctx->panel.prepare_prev_first = true;
+
+    ctx->panel.backlight = k6_38_0e_0b_dsc_create_backlight(dsi);
+    if (IS_ERR(ctx->panel.backlight))
+        return dev_err_probe(dev, PTR_ERR(ctx->panel.backlight),
+            "Failed to create backlight\n");
+
+    drm_panel_add(&ctx->panel);
+
+    /* This panel only supports DSC; unconditionally enable it */
+    dsi->dsc = &ctx->dsc;
+
+    ctx->dsc.dsc_version_major = 1;
+    ctx->dsc.dsc_version_minor = 1;
+
+    /* TODO: Pass slice_per_pkt = 2 */
+    ctx->dsc.slice_height = 40;
+    ctx->dsc.slice_width = 540;
+    /*
+     * TODO: hdisplay should be read from the selected mode once
+     * it is passed back to drm_panel (in prepare?)
+     */
+    WARN_ON(1080 % ctx->dsc.slice_width);
+    ctx->dsc.slice_count = 1080 / ctx->dsc.slice_width;
+    ctx->dsc.bits_per_component = 8;
+    ctx->dsc.bits_per_pixel = 8 << 4; /* 4 fractional bits */
+    ctx->dsc.block_pred_enable = true;
+
+    ret = mipi_dsi_attach(dsi);
```

Comment:
If you switch to devm_mipi_dsi_attach() here you can drop the detaching code in the _remove function. It will detach for you automatically